pkg/proxy/runner/bounded_frequency_runner.go

```diff
@@ -21,188 +21,143 @@ import (
 }
 
 // Make an instance with dependencies injected.
-func construct(name string, fn func(), minInterval, maxInterval time.Duration, burstRuns int, timer timer) *BoundedFrequencyRunner {
+func construct(name string, fn func(), minInterval, retryInterval, maxInterval time.Duration, clock clock.Clock) *BoundedFrequencyRunner {
 	if maxInterval < minInterval {
 		panic(fmt.Sprintf("%s: maxInterval (%v) must be >= minInterval (%v)", name, maxInterval, minInterval))
 	}
-	if timer == nil {
-		panic(fmt.Sprintf("%s: timer must be non-nil", name))
+	if clock == nil {
+		panic(fmt.Sprintf("%s: clock must be non-nil", name))
 	}
 
 	bfr := &BoundedFrequencyRunner{
-		name:        name,
-		fn:          fn,
-		minInterval: minInterval,
-		maxInterval: maxInterval,
-		run:         make(chan struct{}, 1),
-		retry:       make(chan struct{}, 1),
-		timer:       timer,
-	}
-	if minInterval == 0 {
-		bfr.limiter = nullLimiter{}
-	} else {
-		// allow burst updates in short succession
-		qps := float32(time.Second) / float32(minInterval)
-		bfr.limiter = flowcontrol.NewTokenBucketRateLimiterWithClock(qps, burstRuns, timer)
+		name:          name,
+		fn:            fn,
+		minInterval:   minInterval,
+		retryInterval: retryInterval,
+		maxInterval:   maxInterval,
+		run:           make(chan struct{}, 1),
+		ready:         make(chan struct{}), // used to synchronize tests
+		clock:         clock,
 	}
+
 	return bfr
 }
 
 // Loop handles the periodic timer and run requests.  This is expected to be
 // called as a goroutine.
 func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
-	klog.V(3).Infof("%s Loop running", bfr.name)
-	bfr.timer.Reset(bfr.maxInterval)
+	klog.V(3).InfoS("Loop running", "name", bfr.name)
+	defer close(bfr.run)
+
+	// retryIntervalTimer is only started after a Retry() call
+	// so it starts stopped.
+	bfr.retryIntervalTimer = bfr.clock.NewTimer(bfr.retryInterval)
+	bfr.retryIntervalTimer.Stop()
+	defer bfr.retryIntervalTimer.Stop()
+
+	bfr.minIntervalTimer = bfr.clock.NewTimer(bfr.minInterval)
+	defer bfr.minIntervalTimer.Stop()
+
+	bfr.maxIntervalTimer = bfr.clock.NewTimer(bfr.maxInterval)
+	defer bfr.maxIntervalTimer.Stop()
+
+	// Signal the loop is ready
+	close(bfr.ready)
+
 	for {
 		select {
 		case <-stop:
-			bfr.stop()
-			klog.V(3).Infof("%s Loop stopping", bfr.name)
+			klog.V(3).InfoS("Loop stopping", "name", bfr.name)
 			return
-		case <-bfr.timer.C():
-			bfr.tryRun()
+		case <-bfr.maxIntervalTimer.C():
+		case <-bfr.retryIntervalTimer.C():
 		case <-bfr.run:
-			bfr.tryRun()
-		case <-bfr.retry:
-			bfr.doRetry()
+		}
+
+		func() {
+			// stop the timers here to allow the tests using the fake clock to synchronize
+			// with the fakeClock.HasWaiters() method. The timers are reset after the function
+			// is executed.
+			bfr.minIntervalTimer.Stop()
+			bfr.maxIntervalTimer.Stop()
+
+			// do not allow to reset the retryInterval during the execution
+			bfr.retryMu.Lock()
+			defer bfr.retryMu.Unlock()
```

Comment:
The only time we ever call Retry() is from within the execution of fn().

Reply:
As an alternative API, we could get rid of Retry() and have `fn()` return a value indicating if it wanted a retry

Reply:
that sounds an interesting idea

Reply:
@danwinship see https://github.com/kubernetes/kubernetes/pull/131502/commits/b6f091534ff20dddf236baf132f9af5242f78e3a for strawman proposal it looks pretty neat and simple, also the diff is very small if we use a named variable on the proxies , If you give me ack on that I will implement it that way in a new commit